sys/hoon.hoon

```diff
@@ -1335,10 +1335,26 @@
     ?~  b
       a
     $(b t.b, a (put i.b))
+  ::  +has: does :b exist in :a?
   ::
-  ++  has                                               ::  b exists in a check
+  ++  has
     ~/  %has
-    |*  b/*
+    |*  b=*
+    ^-  ?
+    ::  wrap extracted item type in a unit because bunting fails
+    ::
+    ::    If we used the real item type of _?^(a n.a !!) as the sample type,
+    ::    then hoon would bunt it to create the default sample for the gate.
+    ::
+    ::    However, bunting that expression fails if :a is ~. If we wrap it
+    ::    in a unit, the bunted unit doesn't include the bunted item type.
+    ::
+    ::    This way we can ensure type safety of :b without needing to perform
+    ::    this failing bunt. It's a hack.
+    ::
+    %.  [~ b]
+    |=  b=(unit _?>(?=(^ a) n.a))
+    =>  .(b ?>(?=(^ b) u.b))
```

Comment:
Yikes!

It seems like all the other arms of `+in` or `+by` that use the `_?>(?=(^ a) n.a)` pattern have this same problem. But is it really a problem?  Do we need to be able to default them?

Reply:
It's definitely a problem that a bunch of our standard library doesn't enforce type safety. Not all those arms have this problem. Something mysterious happens when the `_?>(?=(^ a) n.a)` is inside the sample definition for a wet gate...

Reply:
I was thinking of `+gas:in`, but that sample is a list and can be defaulted without issue. Something like this pattern is probably necessary for the arms of a treap engine that operate directly on individual items. It would certainly be nicer if it could be more encapsulated somehow.

Not sure what you mean about the wet gate sample -- why would it need to be wet if you have a specific structure like that?

Reply:
I don't know! Seems like samples of wet gates don't get bunted or something.

Reply:
Or maybe wetness has nothing to do with it. The assertion in `+turn` probably only works because it works on a list. Please disregard my earlier comments ...